[28기 김정현][Refactor] 분리 가능한 useEffect 동작 hook으로 분리.

KimJeongHyun opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- Navbar의 햄버거 버튼을 클릭했을 때, 히든 메뉴창이 점멸합니다.
- 메뉴창에서 검색하거나, 장바구니 페이지로 이동하거나, 목록 페이지로 이동할 수 있습니다.
- 메뉴창에서 수직 리스트바를 스크롤할 수 있고, 수평 이벤트 페이지를 스크롤할 수 있습니다.
<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 햄버거 버튼 클릭시 메뉴 점멸
- 햄버거 버튼을 클릭하면 isShow state의 boolean을 토글합니다.
- 해당 값이 true일 때 히든메뉴 컴포넌트가 그려집니다.
- 히든 메뉴 컴포넌트가 그려지면 스크롤바를 가리는 커스텀 훅이 동작합니다.
- isShow를 감지하면, setTimeout을 이용해 render가 늦게 되도록 트릭을 사용합니다.
- 이를 통해 transition 효과를 user에게 보여줄 수 있습니다.

2. 히든 메뉴 수직/수평 스크롤 리스트
- mousewheel event를 통해 수직/수평 스크롤이 가능합니다.
- 수평의 경우 : translateX를 이용해 overflow 영역을 확인할 수 있습니다. 
 - 스크롤 한 번당 각 데이터 1개만큼의 width만큼을 translate합니다. 0 이상이거나, overflow 영역의 offsetWidth이하에서 스크롤됩니다.
- 수직의 경우 : translateY를 이용해 overflow 영역을 확인할 수 있습니다.
 - 스크롤 영역이 420-300 = 120 이하라면, 스크롤되지 않습니다. 여러 경우를 테스트했을 때, 해당 식이 최소한 자연스러운 것으로 확인되었습니다.
 - 데이터가 6개 이상인 경우, 7번째부터 스크롤되어야하므로 스크롤 한 번당 (전체 길이 - 고정된 6개 만큼의 길이) 만큼을 기준으로 잡아 translateY를 더하거나 빼는 것으로 제어됩니다. 다만 데이터가 6개 미만일 때 스크롤 영역이 발생하는 경우가 있으므로, 전체길이 - 고정된 6개 만큼의 길이는 절댓값으로 제어합니다.

3. 로그인, 회원가입, 로그아웃
- 로그인 버튼을 누르면 로그인 페이지로 이동합니다.
- 회원가입 버튼을 누르면 회원가입 페이지로 이동합니다.
- 로그아웃 버튼을 누르면 토큰이 삭제되며, 메인으로 이동합니다.
 - 토큰 유무에 따라 navhead에서 '환영합니다! 로그아웃' 또는 '로그인 회원가입' 버튼을 출력합니다.

4. 목록 페이지
- 지금은 바로 productList로 라우팅되지만, 선택한 카테고리에 따라 다른 endPoint와 쿼리스트링을 조합한 URL로 navigate합니다.
- 쿼리생성과 관련된 함수를 customlib 폴더에 보관하여 다른 페이지에서도 사용가능하도록 했습니다.

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- useLayoutEffect를 사용해보았습니다. 찰나에 스크롤바가 나왔다가 들어가는 부분이 영 맘에 들지 않았는데, 생명주기 메서드 중 Paint 이전에 기능하는 훅을 발견하여 공부해보았고, 결과가 좋습니다. 무턱대고 사용하면 안되겠지만, 바로 유저 눈 앞에 보여줘야하는 연산될 요소가 있다면 적극적으로 채용해도 괜찮겠다는 생각을 했습니다.
- 수직/수평 리스트 스크롤링을 구현할 때 그림을 그리지 않고 하다가 시행착오를 많이 겪었습니다. 코드를 짜기 전에 미리 로직을 글로 작성한 뒤 그것을 코드로 옮겨야겠다고 생각하게 되었습니다.

<br />

## :: 기타 질문 및 특이 사항

## 구현 화면

![hiddenMenuFinal](https://user-images.githubusercontent.com/23470125/148035131-aeac6e43-4660-4e68-b618-6d7d604b960f.gif)

KimJeongHyun committed:
[Add] NavHiddenMenu.js 레이아웃 구현 완료
- Nav 햄버거 버튼 클릭시 히든메뉴 점멸. X 버튼을 누르면 닫아짐.
- 히든 메뉴 켜지면 스크롤바를 비활성화하고, 꺼지면 다시 활성화함.
- 장바구니 / 닫기 버튼 및 검색바 부분 레이아웃 구현.

KimJeongHyun committed:
[Add] 스크롤바 Hide 커스텀 훅으로 변경
- useLayoutEffect 사용
- 어떤 페이지든 modal 호출시 스크롤바가 없어질 수 있도록.

saengmotmi commented:
@KimJeongHyun 리뷰 요청하시는 PR인가요? 라벨이 안달려 있어서요

KimJeongHyun commented:
@saengmotmi 

엇 추가기능구현 라벨 어제 붙였습니다. 혹시 안보이실 수 있을 것 같아서 label 다시 뗐다 붙였습니다!

오늘 내로 동작 완료될 예정이고, 질문거리 들고 찾아뵙겠습니다 😇

saengmotmi commented:
@KimJeongHyun 저는 리뷰 요청 라벨 여쭤본거였습니다 ㅎㅎ 그러면 일단 리뷰 안하고 패스하겠습니다!

KimJeongHyun committed:
[Add] NavHiddenMenu.js 동작 구현 완료. 리팩토링 진행예정

KimJeongHyun committed:
[Fix] sider issue 해결

KimJeongHyun committed:
[Fix] sider issue 해결

KimJeongHyun committed:
[Fix] sider issue 해결

KimJeongHyun committed:
[Modify] NavHiddenMenu.js 가독성을 위한 컴포넌트 분리 진행
- 수직/수평 컨텐츠 부분 분기
- scss 분기
- NavHiddenVertical -> VerticalItems 로 네이밍 변경

KimJeongHyun committed:
[Fix] sider issue 해결

KimJeongHyun committed:
[Modify] main merge지 라우팅 대비 및 히든 메뉴 수정 등
- a태그에  라우팅 경로 명시
- 쿼리스트링은 가칭. 해당 페이지 담당자와 url 논의 후 확정 예정.
- 히든 메뉴 수직 스크롤바 로직 소폭 수정.
- 카테고리 2개 (워커, 운동화) 대비 목업데이터 수정
- JWT 토큰 이름 확인 필요함
- 로그아웃 잘 구현되는지 확인 필요함

saengmotmi commented on src/hooks/useScrollHidden.js:
```diff
@@ -0,0 +1,9 @@
+import { useLayoutEffect } from 'react';
+
+export default function useScrollHidden() {
+  useLayoutEffect(() => {
```
👍 👍 👍

saengmotmi commented on src/components/Nav/NavHiddenVertical/VerticalItems.js:
```diff
@@ -0,0 +1,46 @@
+    setIsSelected(!isSelected);
+  };
+
+  const goToList = event => {
+    const clickedInfo = event.target.getAttribute('name').split('_');
+
+    const listCategory = item.name;
+    const detailCategory = clickedInfo[0];
+    const listId = clickedInfo[1];
```
```suggestion
  const goToList = event => {
    const [detailCategory, listId] = event.target.getAttribute('name').split('_');
    const listCategory = item.name;
```
이렇게도 가능하겠네요

saengmotmi commented on src/components/Nav/NavHiddenVertical/VerticalContent.js:
```diff
@@ -0,0 +1,77 @@
+
+      const moveUp = () => {
+        setVerticalPos(current => {
+          return current - scrollEach > 0 ? current - scrollEach : 0;
```
변수명으로 붙여주면 좋을거 같아요!

saengmotmi commented on src/components/Nav/NavHiddenVertical/VerticalContent.js:
```diff
@@ -0,0 +1,77 @@
+
+import './VerticalContent.scss';
+
+const verticalListObject = verticalListMockData;
```
다시 다른 변수에 옮겨담아주는 이유가 있나요?

saengmotmi commented on src/components/Nav/Nav.js:
```diff
@@ -1,48 +1,91 @@
+        <section className={`navUpper ${visibleValue ? '' : 'hide'}`}>
+          <div className="navUpperFixedWidth">
+            <div className="navUpperBtns">
+              {isValidUser === null ? (
```
```suggestion
              {!isValidUser ? (
```
그냥 이렇게 표현해보면 어떨까요?

saengmotmi commented on src/components/Nav/Nav.js:
```diff
@@ -1,48 +1,91 @@
-              <input placeholder="검색어를 입력해주세요" />
-              <FontAwesomeIcon className="fontAwesomeIcon" icon={faSearch} />
+        </section>
+        <section className={`navBottom ${visibleValue ? '' : 'hide'}`}>
```
```suggestion
        <section className={cn('navBottom', cond(visibleValue, 'hide'))}>
```
시간 되실 때 이런 식으로 쓰일 수 있도록 cn, cond 함수 구현해보시겠어요?

saengmotmi reviewed (commented):
리뷰 확인해주세요~

KimJeongHyun committed:
[Add] SetClassCondition 함수 생성
- 이 함수는 기본 className, 토글용 state, 더해질 className, 3항 연산자의
  방향을 받습니다.
- 3항 연산자의 방향은 기본값이 1로 세팅됩니다.
- 기본값인 경우, state가 true이면 className이 더해집니다.
- 0이 들어오는 경우, state가 false이면 className이 더해집니다.
- 현재 Nav branch에 전부 적용된 상태이며, 타 branch에서 사용가능하도록
  별도의 customlib 폴더에 저장했습니다.

KimJeongHyun commented:
@saengmotmi

리뷰 전부 반영은 되었으나 라우팅 정리 후 리뷰 요청 라벨 붙이겠습니다!

KimJeongHyun committed:
[Modify] localStorage 토큰 이름 변경

KimJeongHyun committed:
[Add] queryString URL 생성기 작성
- SetQueryString(endPoint,condition)
- config로부터 BASE_URL을 가져온다.
- 각 페이지에서 필요한 endPoint를 설정해두고
- 필터를 거는 경우, 해당 필터에 설정된 key이름으로 이루어진
  categories배열과 각각에 대응되는 value배열을 SetQueryCondition.js를
  통해 condition 배열을 받는다.
- 이후, SetQueryString(endPoint,condition)을 걸면 BASE_URL과
  queryString이 적절히 조합되어 반환되며, 이를 바로 navigate하면 된다.

KimJeongHyun committed:
[Refactor] 분리 가능한 useEffect 동작 hook으로 분리.

saengmotmi commented on src/components/customlib/SetClassCondition.js:
```diff
@@ -0,0 +1,18 @@
+export default function SetClassCondition(
+  className,
+  stateValue,
+  addedClassName,
+  addedClassPos
+) {
+  return `${className} ${cond(stateValue, addedClassName, addedClassPos)}`;
+}
+
+function cond(stateValue, addedClassName, addedClassPos = 1) {
+  return addedClassPos
+    ? stateValue
+      ? addedClassName
+      : ''
+    : stateValue
+    ? ''
+    : addedClassName;
+}
```
```typescript
export const cn = (...args: string[]) => args.join(' ');
export const cond = (cond: boolean, className: string) => (cond ? className : '');
```
```js
className={cn('hi', cond(isOpen, 'selected'))}
```
참고해보세요!
그리고 함수 이름은 소문자로 시작해야 합니다

KimJeongHyun replied:
@saengmotmi  아~ 이해됐습니다! 감사합니다

saengmotmi commented on src/components/customlib/SetQueryCondition.js:
```diff
@@ -0,0 +1,12 @@
+export default function SetQueryCondition(categories, items) {
```
이 함수는 사용처가 어디인가요?

KimJeongHyun replied:
@saengmotmi 

SetQueryString(endPoint, condition) 에서 condition은 객체로 구성된 배열입니다.

```
[
 {
  category:'shoecategory',
  value:'워커'
 },
 {
  category:'gender',
  value:'male'
 }
]
```
예를 들면 condition은 위와 같은 형태로 들어와야합니다.

![image](https://user-images.githubusercontent.com/23470125/148317492-a7f8354b-5a39-476e-8191-8f33111e0393.png)

위의 사진에서 정렬의 인기순, 스타일의 오리지널을 선택하면

```
categories =['sort','style']
value=['인기순','오리지널']
```

해당 컴포넌트에서 각각을 이렇게 배열에 집어넣고

이걸 SetQueryCondition에 던지면 그에 맞는 객체 배열을 생성해주고

```
SetQueryString('product',SetQueryCondition(categories:array, items:array))
```
이런 식으로 사용합니다만

어...이걸 왜 굳이 이렇게 했는지 어제의 제가 이해가 안되네요ㅋㅋㅋㅋ

다음 리팩토링 때 지우겠습니다 ㅜㅜ

saengmotmi reviewed (commented):
일단 머지하겠습니다 코멘트 확인해주세요!
